File: backend/src/main/java/com/example/soundwave/service/SongService.java
package com.example.soundwave.service;

import com.example.soundwave.dto.musicbrainz.SongDto;
import java.util.List;

/**
 * Service interface for artist-related business logic
 *
 * This layer abstracts where artist data comes from
 * (MusicBrainz, db, cache...)
 */

public interface SongService {

    /**
     * Returns a list of random songs
     *
     * @return list of SongDto
     */

    List<SongDto> getRandomSongs();


    /**
     * Search for songs by name
     *
     * @param query search keyword
     * @return list of matching songs
     */

    List<SongDto> searchSongs(String query);


    /**
     * Fetch a single song by Id
     *
     * @param songId MusicBrainz song Id
     * @return SongDto
     */

    SongDto getSongById(String songId);


}

File: frontend/src/services/songService.js
const API_BASE_URL = "http://localhost:8080/songs";

export async function getRandomSongs() {
  const response = await fetch(`${API_BASE_URL}/random`);
  if (!response.ok) {
    throw new Error("Failed to fetch songs");
  }
  return response.json();
}

export async function searchSongs(query) {
  const response = await fetch(
    `${API_BASE_URL}/search?q=${encodeURIComponent(query)}`
  );
  if (!response.ok) {
    throw new Error("Failed to search songs");
  }
  return response.json();
}

export async function getSongById(songId) {
  const response = await fetch(`${API_BASE_URL}/${encodeURIComponent(songId)}`);
  if (!response.ok) {
    throw new Error("Failed to fetch song");
  }
  return response.json();
}
